Bottom: 148c2876f155505ac4278ee270c20035562ece2f
Top:    a120429eda084ccc9a0974e30fe0bccff64d59d4
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2016-01-06 16:47:37 +0000

Test some polygon stuff


---

diff --git a/src/polygon1.c b/src/polygon1.c
index 4996842..d32a501 100644
--- a/src/polygon1.c
+++ b/src/polygon1.c
@@ -2001,6 +2001,7 @@ M_POLYAREA_update_primary (jmp_buf * e, POLYAREA ** pieces,
   int del_inside = 0;
   int del_outside = 0;
   int finished;
+  int everything_holes;
 
   if (a == NULL)
     return;
@@ -2038,6 +2039,7 @@ M_POLYAREA_update_primary (jmp_buf * e, POLYAREA ** pieces,
 	{
 	  anext = a->f;
 	  finished = (anext == *pieces);
+	  everything_holes = 0;
 
 	  /* Test the outer contour first, as we may need to remove all children */
 
@@ -2059,6 +2061,9 @@ M_POLYAREA_update_primary (jmp_buf * e, POLYAREA ** pieces,
 	      /* a->contours now points to the remaining holes */
 	      poly_DelContour (&curc);
 
+	      everything_holes = 1;
+
+#if 1
 	      if (a->contours != NULL)
 		{
 		  /* Find the end of the list of holes */
@@ -2076,6 +2081,7 @@ M_POLYAREA_update_primary (jmp_buf * e, POLYAREA ** pieces,
 	      poly_Free (&a);	/* NB: Sets a to NULL */
 
 	      continue;
+#endif
 	    }
 
 	  /* Loop whilst we find INSIDE contours to delete */
@@ -2118,6 +2124,25 @@ M_POLYAREA_update_primary (jmp_buf * e, POLYAREA ** pieces,
 	    }
 	  /* End check for deleted holes */
 
+	  if (everything_holes)
+	    {
+	      if (a->contours != NULL)
+		{
+		  /* Find the end of the list of holes */
+		  curc = a->contours;
+		  while (curc->next != NULL)
+		    curc = curc->next;
+
+		  /* Take the holes and prepend to the holes queue */
+		  curc->next = *holes;
+		  *holes = a->contours;
+		  a->contours = NULL;
+		}
+
+	      remove_polyarea (pieces, a);
+	      poly_Free (&a);	/* NB: Sets a to NULL */
+	    }
+
 	  /* If we deleted all the pieces of the polyarea, *pieces is NULL */
 	}
       while ((a = anext), *pieces != NULL && !finished);
